src/material-experimental/mdc-slider/slider.ts

```diff
@@ -773,9 +772,53 @@ export class MatSlider extends _MatSliderMixinBase
 class SliderAdapter implements MDCSliderAdapter {
 
   /** The global change listener subscription used to handle change events on the slider inputs. */
-  changeSubscription: Subscription;
+  private _changeSubscription: Subscription;
 
-  constructor(private readonly _delegate: MatSlider) {}
+  /** The global input listener subscription used to handle input events on the slider inputs. */
+  private _inputSubscription: Subscription;
+
+  /** The MDC Foundations handler function for input change events. */
+  private _inputChangeEventHandlers = new Map<Thumb, SpecificEventListener<EventType>>();
+
+  constructor(private readonly _delegate: MatSlider) {
+    this._changeSubscription = this._subscribeToInputEvents('change',
+      (event: Event, thumbPosition: Thumb) => {
+        const handler = this._inputChangeEventHandlers.get(thumbPosition)!;
+        handler(event);
+      }
+    );
+    this._inputSubscription = this._subscribeToInputEvents('input');
+  }
+
+  /**
+   * Handles "change" and "input" events on the slider inputs.
+   *
+   * Exposes a callback to allow the MDC Foundations "change" event handler to be called for "real"
+   * events.
+   *
+   * ** IMPORTANT NOTE **
+   *
+   * We block all "real" change and input events and emit fake events from #emitChangeEvent and
+   * #emitInputEvent, instead. We do this because interacting with the MDC slider won't trigger all
+   * of the correct change and input events, but it will call #emitChangeEvent and #emitInputEvent
+   * at the correct times. This allows users to listen for these events directly on the slider
+   * input as they would with a native range input.
+   */
+  private _subscribeToInputEvents
+    (type: 'change'|'input', callback?: (event: Event, thumbPosition: Thumb) => void) {
+      return this._delegate._globalChangeAndInputListener.listen(type, (event: Event) => {
+        const targetIsEndInput = event.target === this._delegate._getInputElement(Thumb.END);
```

Comment:
maybe extract the listener to a named private function for readability

Reply:
Two questions on this:

1. What name do you have in mind for the new private function?

2. Do you mean something like this where the entire listener is moved to a new function:

```ts
  private _subscribeToInputEvents
    (type: 'change'|'input', callback?: (event: Event, thumbPosition: Thumb) => void) {
      return this._delegate._globalChangeAndInputListener.listen(type, (event: Event) => {
        this._handleGlobalChangeOrInputEvent(event, callback, thumbPosition);
      });
  }
  
  private _handleGlobalChangeOrInputEvent
    (event: Event, callback?: (event: Event, thumbPosition: Thumb) => void) {
      const targetIsEndInput = event.target === this._delegate._getInputElement(Thumb.END);
      const targetIsStartInput = this._delegate._isRange() &&
        event.target === this._delegate._getInputElement(Thumb.START);
      if (targetIsEndInput || targetIsStartInput) {
        if ((event as any).isFake) { return; }
        event.stopImmediatePropagation();
        if (callback) {
          callback(event, targetIsStartInput ? Thumb.START : Thumb.END);
        }
      }
  }
```

Or like this:

```ts
  private _subscribeToInputEvents
    (type: 'change'|'input', callback?: (event: Event, thumbPosition: Thumb) => void) {
      return this._delegate._globalChangeAndInputListener.listen(type, (event: Event) => {
        const targetIsEndInput = event.target === this._delegate._getInputElement(Thumb.END);
        const targetIsStartInput = this._delegate._isRange() &&
          event.target === this._delegate._getInputElement(Thumb.START);
        if (targetIsEndInput || targetIsStartInput) {
          this._handleInputOrChangeEvent(event, callback, thumbPosition);
        }
      });
  }
  
  private _handleInputOrChangeEvent
    (event: Event, callback?: (event: Event, thumbPosition: Thumb) => void) {
      if ((event as any).isFake) { return; }
      event.stopImmediatePropagation();
      if (callback) {
        callback(event, targetIsStartInput ? Thumb.START : Thumb.END);
      }
  }
  ```